### DISCUSSION TITLE
FR: create links while in 'Link map' mode

### DISCUSSION THREAD

**Discussion by quesada** [zim-desktop-wiki/zim-desktop-wiki#1892] [Ideas]
From UX POW, It could be as simple as 'drag from a node (arrow appears), touch another node'.
This requires the graph display to be iterative/interactive, which is is not something zim has right now if I undestand it right: graphviz is called in batch mode.

So a very, very 'long term' feature request if at all interesting to the community.

**Comment by quesada** [comment]
https://www.graphviz.org/resources/ lists quite a few dot file interactive visualizers, FWIW. The plugin uses xdot so far, which is static. @jaap-karssenberg maybe a good way to go would be to use the config window in 'link map' plugin to select between xdot and any other interactive dot format GUI?

How to transfer the changes from the GUI to actual links between notes is a bigger task, that's for sure :) But even if I have to do these by hand it'd be worth it for me.

**Reply by quesada** [reply]
This one looks promising and is multiplatform: https://github.com/ArsMasiuk/qvge.
EDIT: that editor is simple and really well done. After playing with it for a while, I think this is a winner. Integrating zim with qvge would position zim as the best network feature set of all notetakng apps.

Is the dot file that the plugin generates available? Maybe an option to keep it somewhere would be a good starting place. And a simple python script to turn a dot file into zim notes would be a second step. I don't think I can attempt this, but it sounds really tempting!

**Reply by jaap-karssenberg** [reply]
I'm open to improvements, but not an area I will spend time in myself anytime soon. Afraid I have other things on the backlog for now.

**Reply by introt** [reply]
I've made a PoC for exports (#1955), discussing it further here: https://github.com/zim-desktop-wiki/zim-desktop-wiki/discussions/1890#discussioncomment-2246517. Adding a menu item which exports the graph for the current page as an attachment & opens it in whatever editor is easy, creating new notes from a dot file sounds doable (assuming they don't need to be pretty; graphviz includes graph traversal tools that can be used for a MVP), but a roundtrip is going to be pretty hardcore.

**Comment by introt** [comment]
Wish there was an easy way to merge discussions, I feel like my replies are getting as scattered as my thoughts :)

Anyway, I've been thinking for some time about extending the Zim webserver via plugins to allow some remote interactions like server-side text search[^search] and appending text to pages[^append] via APIs that could be called by either JS on the pages themselves or some third-party client.

While looking for an open source graph visualizer I realized that a web tech based one would be the easiest to integrate with both what we have right now as well as the aforementioned future functionality, whenever it gets implemented.

[Cytoscape.js](https://js.cytoscape.org/) seems to be the popular one. Looking at their [demos](https://github.com/cytoscape/cytoscape.js/tree/5fc9b2a1fa3b8c9929f0e2e7b331d5bf3987f341/documentation/demos/linkout-example), the data format looks like this:

```javascript
  elements: {
    nodes: [
      { data: { id: 'desktop', name: 'Cytoscape', href: 'http://cytoscape.org' } },
      { data: { id: 'js', name: 'Cytoscape.js', href: 'http://js.cytoscape.org' } }
    ],
    edges: [
      { data: { source: 'desktop', target: 'js' } }
    ]
  },
```

We can achieve this with minimal modifications to the current code, but that's not the best part - licensing aside, since the webserver already serves static pages built on-demand server-side using templates, it should be trivial to turn one of the Cytoscape demos into a template for serving static link maps which can be included in offline HTML exports as well. Additional web server/link graph functionalities could then go hand-in-hand without needing to duplicate work between web (API) development and link map integration.

[^search]: Blocked by #64/search API rework; client-side search using JS is way too slow for big notebooks.
[^append]: Likely via extending [the append plugin](https://github.com/Osndok/zim-plugin-append). I'd like to keep the server stuff read-only and keep the master copy on my PC, so sending patches generated by the server via email to be merged at home might be the way to go ;)

**Reply by introt** [reply]
By "appending text to pages via an API" I'm thinking of having a spartan js-free html form at the end of the page which sends a POST request containing the appendage to the server which adds it to the end of the page after a horizontal bar (perhaps followed by some tag as well as the date), resulting in something (visually) reminiscent of the wikis of the past.
